src/NUnitFramework/framework/Constraints/CollectionEquivalentConstraint.cs

```diff
@@ -67,13 +73,36 @@ public override string Description
         /// <returns></returns>
         protected override bool Matches(IEnumerable actual)
         {
-            // This is just an optimization
-            if (_expected is ICollection && actual is ICollection)
-                if (((ICollection)actual).Count != ((ICollection)_expected).Count)
-                    return false;
+            //Store the tally so the collection is only iterated over once.
+            _lastPerformedTally = Tally(_expected);
+
+            return _lastPerformedTally.TryRemove(actual) && _lastPerformedTally.MissingItems.Count == 0;
+        }
+
+        /// <summary>
+        /// Test whether the collection is equivalent to the expected.
+        /// </summary>
+        /// <typeparam name="TActual">
```

Comment:
Picky - can we put all these `<param>`/`<returns>` etc. docs into single lines - to match the rest of the code-base?